src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyInfoDictionaryValueList.cs

```diff
@@ -65,8 +63,10 @@ public void FinishEditingAndMakeReadOnly(Type parentType)
 
             // We do not know if any of the keys needs to be updated therefore we need to re-create cache
             _parent.Clear();
+            // Perform a final sort of properties before rebuilding the cache
+            _items.StableSortByKey(static prop => prop.Order);
```

Comment:
make sure there is a test which never touches Properties until Serialize is called and then checks they're properly sorted. I think before this change FinishAndMakeReadonly was never called in some cases, make sure to cover that case

Reply:
I'm planning on adding tests verifying this as well as stability, but you can't really do this reliability without exposing the `Order` property.